Support `defer()` in `source()` and `knit()`

lionel- opened the pull request:
Following Jenny's suggestion about "document scope", and Hadley's suggested implementation, `defer()` within `source()` and `knitr::knit()` now detects the relevant frame to register deferred expressions. This way, expressions are executed when `source()` and `knit()` return. In Rmd documents this gives `local_` functions "document scope" rather than "session scope".

The detection of `source()` adds 10 microseconds to every `defer()` call on an empty stack:

```r
bench::mark(
  defer = local(defer(NULL))
)[1:8]

#> # main
#>   expression      min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc
#>   <bch:expr> <bch:tm> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>
#> 1 defer        17.2µs 18.7µs    46136.      280B     41.6  9991     9

#> # PR
#>   expression      min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc
#>   <bch:expr> <bch:tm> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>
#> 1 defer        26.7µs 28.6µs    31458.      280B     47.3  9985    15
```

and about 25 on a stack of about 50 frames:

```r
f <- function(n) {
  if (n) {
    f(n - 1)
  } else {
    defer(NULL)
  }
}

bench::mark(
  defer = f(50)
)[1:8]

#> # main
#>   expression      min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc
#>   <bch:expr> <bch:tm> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>
#> 1 defer        25.9µs 27.3µs    32594.        0B     42.4  9987    13

#> # PR
#>   expression      min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc
#>   <bch:expr> <bch:tm> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>
#> 1 defer        51.5µs 53.8µs    18078.     111KB     50.0  8316    23
```

hadley commented on tests/testthat/test-defer.R:
```diff
@@ -104,3 +104,41 @@ test_that("defer executes all handlers even if there is an error in one of them"
+
+  out <- NULL
+  cat(file = file, "
+    out <<- c(out, '1')
```
Is it worth adding a nested `source()` just to double check that it's attached to the right source frame? Or maybe that's already implied because maybe testthat uses source`?

lionel- replied:
Good idea.

hadley commented on R/compat-defer.R:
```diff
@@ -66,21 +75,105 @@ setup_handlers <- function(envir) {
+  }
+  calls <- as.list(calls)
+
+  if (!is_call(calls[[i - 0]], quote(eval))) {
+    return(NULL)
+  }
+  if (!is_call(calls[[i - 1]], quote(eval))) {
+    return(NULL)
+  }
+  if (!is_call(calls[[i - 2]], quote(withVisible))) {
+    return(NULL)
+  }
```
Is it really worth it to check all these intermediate frames?

lionel- replied:
I think so, otherwise things like `evalq(defer(foo))` within the sourced script will be scoped in `source()` instead of the local frame.

lionel- commented:
Interestingly I struggled with the `source()` unit tests while preparing this PR, it caused a C-level stackoverflow crash. The `R_cycle_detected()` function, called when setting an attribute (in this case the handlers) failed to detect cycles and called itself recursively. I haven't analysed exactly what was happening and I'm not sure what was different about this test. To fix this I changed the handler data structure from `<list expr = , env = >` to `<env expr = , env = >`. This also fixed the infinite recursion in `print()` and `str()` when printing an environment frame with handlers attached.

This will all be better once we can directly use the `on.exit()` stack instead of our own handler list. We will be able to switch when the last R version that doesn't feature the `after` argument is no longer supported.

lionel- commented:
Merging now but feel free to review or comment @jennybc
